_posts/2023-01-24-baking-2d-lighting.markdown

```diff
@@ -0,0 +1,187 @@
+    {
+        rt = new RenderTexture(512, 512, 0, RenderTextureFormat.ARGBFloat)
+        {
+            filterMode = FilterMode.Bilinear,
+            antiAliasing = 4
+        };
+        rt.Create();
+        result = new RenderTexture(rt.width, rt.height, rt.depth, RenderTextureFormat.DefaultHDR);
+        result.Create();
+        cam.targetTexture = rt;
+    }
+```
+
+All this block of code does is initalize two identical render textures (the only difference being their format, which I found through trial & error with the shader I'll get into later). It tehjen assigns our "Baking Camera" to render onto one of the render textures. Now that the camera is setup, I can start "taking pictures".
+
+```csharp
+    private IEnumerator LightAnim()
+    {
+        while (lightBakingQueue.Count > 0)
+        {
+            //I created a queue of lights that needed to be baked by sorting lights by distance
+            // to the player. That way lights closer to the player get baked faster
+            currentLight = lightBakingQueue.Dequeue();
+            currentLight.ToggleLight(true);
+            //I had to delay the next section because sometimes the light baking would fail
+            // my theory is that it takes a frame for Unity's lighting system to start working?
+            yield return null;
+            int iterations = 0;
+            while (iterations < maxIterations)
+            {
+                iterations++;
+                //These iterations are what create the soft shadows. By randomizing the light position every iteration,
+                // we are simulating a light source that is larger than a singular point of light. I started with 10
+                // iterations per light before realizing that 3 was more than enough to create soft shadows
+                currentLight.RandomizeLightPos();
+                cam.Render();
+                // I'll get into the shader being used here later
+                Graphics.Blit(rt, result, blitMat);
+                yield return null;
+            }
+            currentLight.ToggleLight(false);
+            currentLight = null;
+        }
+    }
+```
+
+To accomplish runtime baking that didn't cause massive lag, I run the baking system in the coroutine shown above. This ensures that each light takes `maxIterations` frames to bake. In a scene with 10+ lights and 3 iterations per light, it can take over 30 frames to bake the whole scene. Therefore, I placed all the lights in a queue based on distance to the player. That way lights closest to the player will bake sooner, minimizing the player experiencing any weird visuals.
+
+Lets look at what the camera bakes now...
+
+| Iteration 1 | Iteration 2 | Iteration 3 |
+| ---- | ---- | ---- |
+| ![Bake 1](/assets/img/baking-2d-lighting/light_bake_1.png "Light Bake 1") | ![Bake 2](/assets/img/baking-2d-lighting/light_bake_2.png "Light Bake 2") | ![Bake 3](/assets/img/baking-2d-lighting/light_bake_3.png "Light Bake 3") |
```

Comment:
Most of the space in these images is unlit. It may help to crop these to the portion that has lighting.